### PR TITLE
initial draft of supervisor package

### PR DESCRIPTION
**Pull request by rhs**
This work is motivated by a number of separate but related problems we have run into over the course of developing teleproxy, apictl, and kubewatch.

1. Much of the structure of what teleproxy does is to manage distinct goroutines and/or subprocesses in such a way as to maintain connectivity, routing, and dns. Off the top of my head it manages a set of intercepts using the local firewall, an in-process dns server, an in-process routing table, an in-process api server for accessing/modifying said routing table, a kubernetes bridge, a docker bridge, a kubectl port-forward, an ssh port-forward, and a partridge in a pear tree. None of these are super complicated in and of themselves, however they all have various dependency relationships between them and there are some hard cleanup requirements as well (e.g. never leave the firewall in a bad state).

   The way this is currently written spaghetties a lot of this concurrency/cleanup/retry logic around the code, so it's really kinda hard to follow/reason about. The hope is that this supervisor abstraction will allow the code, the dependencies, and the cleanup logic to be expressed in a much clearer way.

2. Teleproxy currently only handles outbound stuff. As we have developed apictl more, it has become clear that we will need a daemon process to manage inbound as well, and this will likely either make the teleproxy concurrency situation worse if we extend it to handle inbound, or introduce another program with similar properties.

3. With kubewatch getting extended to interact with consul in addition to kubernetes, it seems likely we will be creating yet another long running go process that has different moving sub parts (the part that watches kubernetes, the part that watches consul, and the part that assembles/supervises ambassador reconfigs).

So all in all it seems like an abstraction to centralize and simplify how we do this sort of thing seems useful. I think the ultimate goal and true test of this abstraction would be whether we can use this to cleanly define the different workers within a long-running program... e.g. what dependencies each worker has, what it's retry policy is, and then just write a bunch of straightforward business logic to implement each worker itself.

### CODE REVIEW

**Comment by plombardi89** [comment]
I glanced at it, but I don't know enough advanced Go to have any strong opinions about it. All I ask for is good examples if we're going to switch to using this over something else.

File: pkg/supervisor/supervisor.go [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+		}
+		s.changed.Broadcast()
+	}()
+}
+
+type Process struct {
+	Supervisor *Supervisor
+	Worker     *Worker
+	// Used for hard cancel.
+	Context context.Context
```
Comment (LukeShu) [inline pkg/supervisor/supervisor.go]: I'd consider making these private, and exposing them in a read-only way via getter methods, like you did for `shutdown`.

I'm not sure what the use-case is for exposing the `Worker` to the `Work` func.

File: pkg/supervisor/supervisor.go [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+	cancel  context.CancelFunc
+	// Used to signal graceful shutdown.
+	shutdown       chan struct{}
```
Comment (LukeShu) [inline pkg/supervisor/supervisor.go]: The asymmetry of having to select on `process.Shutdown()` for graceful-shutdown and `process.Context.Done()` for not-so-graceful-shutdown seems weird.  Why not use a context for both? Nothing about either of those names suggests which one is graceful and which one is not-so-graceful.

File: pkg/supervisor/supervisor.go [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+}
+
+// launch anything that is safe to launch and not already launched
+func (s *Supervisor) reconcileNormal() {
+OUTER:
+	for _, n := range s.names {
+		w := s.workers[n]
+		if w.process == nil {
+			for _, r := range w.Requires {
+				process := s.workers[r].process
```
Comment (LukeShu) [inline pkg/supervisor/supervisor.go]: This seems like it could be an easy accidental panic (de-reference nil), if you forget to .Supervise() a worker that another worker depends on.

File: pkg/supervisor/supervisor.go [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+}
+
+// Used for graceful shutdown...
+func (p *Process) Shutdown() <-chan struct{} {
+	return p.shutdown
+}
+
+// Used for logging...
+func (p *Process) Log(obj interface{}) {
+	log.Printf("%s: %v", p.Worker.Name, obj)
```
Comment (LukeShu) [inline pkg/supervisor/supervisor.go]: I don't like that this hard-codes the global `log.std` logger. IMO, that would be a blocker to using it in APro.  I would give the supervisor an optional Logger setting that takes an `interface { Printf(format string, v ...interface{}) }`; or maybe even a LoggerFactory that is something like

```go
type Logger interface {
    Printf(format string, v ...interface{})
}
type LoggerFactory interface {
    NewLogger(workerName string) Logger
}
```

File: pkg/supervisor/supervisor.go [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+func (s *Supervisor) Run() []error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	for len(s.workers) > 0 {
+		s.reconcile()
+		// reconcile may delete workers
+		if len(s.workers) > 0 {
+			s.changed.Wait()
+		}
+	}
```
Comment (LukeShu) [inline pkg/supervisor/supervisor.go]: Wouldn't it be cleaner to write this as

```go
s.reconcile()
for len(s.workers) > 0 {
    s.changed.Wait()
    s.reconcile()
}
```

File: pkg/supervisor/supervisor.go [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+	delete(s.workers, worker.Name)
+	var newNames []string
+	for _, name := range s.names {
+		if name == worker.Name {
+			continue
+		} else {
+			newNames = append(newNames, name)
+		}
+	}
+	s.names = newNames
```
[the remark this answers is not in the document]
Comment (rhs) [reply pkg/supervisor/supervisor.go]: I wanted to keep the startup/shutdown order deterministic.

File: pkg/supervisor/supervisor.go (hunk 1 above)
Comment (rhs) [reply pkg/supervisor/supervisor.go]: I've done as you suggested. The use case for exposing the Worker from the Process object was to access the name when logging. Strictly speaking that doesn't need to be public as-is, but with a pluggable logger I think it would actually need to be public.

File: pkg/supervisor/supervisor.go (hunk 2 above)
Comment (rhs) [reply pkg/supervisor/supervisor.go]: This is the last outstanding comment... I agree the names could be better... my rationale for not using a context is that in my head shutdown is not actually a signal to end anything, it's a signal to begin something (i.e. the shutdown process) which could take an indefinite amount of time. This struck me as a bit of a departure from the context use case. This is sorta in line with what the http.Server type does in the std library (they have Server.Shutdown(context)).

How strongly do you feel about this? I have a minor preference for keeping the asymmetry but trying to come up with better naming and adding docs. If you feel strongly or are aware of some similar uses of contexts I'm happy to defer to your judgement.

Comment (LukeShu) [reply pkg/supervisor/supervisor.go]: I don't feel super strongly about this.  But I will note that `http.Server` (and much of the standard library) is a little non-idiomatic in its use of contexts, because contexts were a retrofit in go1.7.

**Comment by ark3** [comment]
I'm trying to understand this but it's going slowly. I shall write out what I've understood so far; please confirm and/or correct.

**Usage**

1. Prepare a supervisor and a set of workers.
2. Launch all the work in the correct order.
3. Block/wait for all the work to finish successfully, or an error condition to arise, in which case the remaining work is interrupted.

**Consequences**

1. Work is automatically launched in the correct order based on the specified dependencies.
2. Work that depends on other work waits until the other work is _ready/running_, not just _started_.
3. All work must be specified before any work happens. You can't have one piece of work compute a set of jobs incrementally and fire them off in the background **using the same supervisor**.

**asdf**

1. Would it be possible to implement the goroutine management separately from the worker state and dependency stuff, in different layers?
2. Would it make sense to use someone else's code? E.g., [Tomb](https://godoc.org/gopkg.in/tomb.v2) does the goroutine stuff (with slightly different semantics).
3. Can we add an API to make a subordinate supervisor so that the use case presented in consequence 3 above is easier?

**Comment by LukeShu** [comment]
>     3. All work must be specified before any work happens. You can't have one piece of work compute a set of jobs incrementally and fire them off in the background **using the same supervisor**.

Yes you can.  It's valid to `.Supervise(w)` more workers after the supervisor has started.    `process.Go(fn)` is even a shortcut for this (`process.Supervisor().Supervise(&Worker{Name: "…", Work: fn})`)

**Comment by rhs** [comment]
The way I would describe the goals are:

1) Supervisor.Run() should give you an easy way to invoke concurrent work that provides a strong guarantee of no goroutines leaked and no errors lost (provided you work through the Supervisor API). This is defined as when Supervisor.Run() exits, there should be no background goroutines, and any errors/panics that may have occurred are reported in the result of Supervisor.Run() (excluding those explicitly ignored via the Retry flag). It shouldn't matter whether you setup the workers in advance and then call Run() or whether some of the workers get setup during Run().

2) On top of the above guarantee there is the ordering guarantees which I think you have described correctly in the startup case... i.e. a worker won't start until it's requirements are ready. This ordering is also respected for shutdown. A worker won't get shutdown until any dependents are already shutdown. There is also the added property that an error in one worker will trigger a shutdown in a way that gives other workers the opportunity to cleanup before exiting... i.e. exit hooks that respect the startup/shutdown order.

Regarding your questions around whether there is a way to layer this... maybe? You could probably create a worker that wraps another worker or something like that and then restarts it or waits until certain preconditions are met to launch it, but I'm having a bit of trouble picturing it right now in a way that seems simpler overall (partly because the motivating use cases seem to want to use all the features at once, and so that's influencing my measurement of "simple").

File: go.mod [hunk 7]
```diff
@@ -36,6 +36,7 @@ require (
 	golang.org/x/crypto v0.0.0-20190123085648-057139ce5d2b // indirect
 	golang.org/x/net v0.0.0-20190125091013-d26f9f9a57f3
 	golang.org/x/oauth2 v0.0.0-20190115181402-5dab4167f31c // indirect
+	golang.org/x/sync v0.0.0-20181221193216-37e7f081c4d4
```
Comment (LukeShu) [inline go.mod]: This doesn't use `x/sync`, does it? Does `go mod tidy` revert this addition?

**Comment by ark3** [comment]
Thanks for answers so far. What's the effective difference between `s.Shutdown()` and `cancel()` as far as the code sees? I'm trying to understand things by reading the tests, but I don't quite get it.

**Comment by rhs** [comment]
The s.Shutdown() option will attempt to do a graceful shutdown. Graceful shutdown respects the ordering implied by Requires. The cancel() option should match idiomatic usage and just "free all resources in no particular order".

As an example, the http package has a Server.Shutdown(ctx) which does a graceful shutdown by waiting for any outstanding requests to finish and what not. This is actually a process that could take an indefinite amount of time (hence why you pass a context into it). If an http server were a worker, I would expect Supervisor.Shutdown to invoke http.Server.Shutdown(Process.Context()).

**Comment by ark3** [comment]
The http package is not a good example for the reason Luke mentioned earlier.

Anyhow, what's the actual difference? What does the code see?

**Comment by rhs** [comment]
I'm not sure I understand your comment. I was trying to give an example of what a graceful shutdown might be. The http.Server.Shutdown(...) example is exactly that... trying to shutdown in a way that does not generate spurious errors and therefore might take a while. The non-idiomatic usage of contexts is really an orthogonal thing. As far as I'm aware there is no common idiom where contexts are used for graceful shutdown, http or otherwise.

From a code perspective the differences are:

Each process gets its own shutdown channel whereas all processes share a single context.
A Supervisor.Shutdown() will trigger a close on the Process.Shutdown() channels in an order that respects the Requires fields of the Workers, i.e. a Worker will not be Shutdown until anything that Requires it has exited.

Canceling the context for a supervisor will end up "simultaneously" canceling everything all at once. If you do cancel the supervisor's context then all the Shutdown() channels are closed simultaneously so if you don't actually care about the difference between hard and soft shutdown then you can just write your worker to pay attention to Shutdown() and the right thing happens.

The expected semantics are that you should try to gracefully shut down when the Shutdown() channel is closed, and you should terminate immediately if the context is canceled. The former is something that is ok to take an indefinite amount of time, e.g. interact with remote resources, etc, the latter is something that needs to be realtime.

**Comment by rhs** [comment]
FYI, I merged this at Phil's request, so he can play with it in the PoC he is working on for kubewatch++, but I'm still considering it an open question how we want to handle this stuff. (As of yet, it's not actually used, so we could roll it back in favor of Tomb or something else if we want.)
